Use `TORCH_CHECK()` instead of throwing `std::runtime_error` in `XLA_CHECK*()` macros' implementation.

ysiraichi opened the pull request:
This PR replaces the XLA debug macro implementation to use PyTorch's `TORCH_CHECK` instead of throwing `std::runtime_error` directly, which makes it so `XLA_CHECK*()` macros show PyTorch C++ stacktrace.

**Key changes:**
- Updated _tf_logging.cpp_ to use `TORCH_CHECK` instead of `std::runtime_error`
- Updated test to match new PyTorch stacktrace format
- Deprecated `ConsumeValue` function in favor of `GetValueOrThrow`
- Moved `THROW_RUNTIME_ERROR_FROM_C10_ERROR` macro to _cpp_test_utils.h_, and added `_IMPL` to its name
    - Both _test_status_common.h_ and _test_debug_macros.cpp_ make use of that macro for throwing the appropriate error
- Add deprecation comment to the `XLA_CHECK*()` macros
    - Use status propagation instead
- Fixed `GetComputationClientOrDie()` (at _runtime.h_) deprecation warning

ysiraichi commented on test/cpp/test_debug_macros.cpp:
```diff
@@ -1,29 +1,44 @@
 
+// Prefix of the C++ stacktrace PyTorch adds to the error message.
+constexpr char kTorchCppStacktracePrefix[] =
+    "Exception raised from operator& at torch_xla/csrc/runtime/tf_logging.cpp:";
```
No. That's [the function](https://github.com/pytorch/xla/blob/43589c03b72d5ee07046886f6eb5212c67ff1ed5/torch_xla/csrc/runtime/tf_logging.cpp#L14) where we call `TORCH_CHECK()`.

zhanyong-wan commented on torch_xla/csrc/runtime/debug_macros.h:
```diff
@@ -30,6 +37,10 @@
 #define XLA_CHECK_GT(a, b) TF_CHECK_GT(a, b)
 
 template <typename T>
+ABSL_DEPRECATED(
```
I don't think we should deprecate this. It has a different use case than GetValueOrThrow().

Sometime we know that the StatusOr must contain a value (e.g. we validated it earlier) and violating this indicates a torch_xla bug. In that case, we do want the failure to crash instead of being reported to the user as a python exception.

ysiraichi replied:
But, in this case, I think we would better use `ABSL_CHECK(status.ok());`. What do you think?

zhanyong-wan replied:
Are you suggesting that instead of
```
a = ConsumeValue(foo);
```
we write
```
ABSL_CHECK(foo.status().ok());
a = foo.value();
```
?

I think the latter is worse:
1. it's more verbose,
2. on failure, it doesn't show the error message in the Status.

zhanyong-wan commented on test/cpp/test_status_common.h:
```diff
@@ -77,36 +78,20 @@ class StatusTest : public testing::TestWithParam<CppStacktracesMode> {
 
 namespace testing {
 
+#define THROW_RUNTIME_ERROR_FROM_C10_ERROR(block) \
```
Can we prefix our macros with `XLA_` to:

- avoid potential clashes with other libraries,
- make it clear that they are our own macros
?

zhanyong-wan commented on test/cpp/test_debug_macros.cpp:
```diff
@@ -1,29 +1,40 @@
 #include "torch_xla/csrc/runtime/debug_macros.h"
 #include "torch_xla/csrc/runtime/env_vars.h"
 
-namespace torch_xla {
```
Why removing this? We don't want to pollute the global namespace.

zhanyong-wan commented on test/cpp/cpp_test_util.h:
```diff
@@ -24,6 +24,36 @@
+//
+// If a `c10::Error` is thrown, this macro will catch it and throw a
+// `std::runtime_error` with backtrace, since `show_backtrace` is true.
+#define XLA_THROW_RUNTIME_ERROR_FROM_C10_ERROR_(block, show_backtrace)         \
```
As commented earlier, I think it's better to remove this macro and just let the tests catch a c10::Error and verify its what() or what_without_backtrace().

ysiraichi replied:
Ah, sorry. I thought it was only for the `test_debug_macros.cpp`.

zhanyong-wan commented on test/cpp/test_status_common.h:
```diff
@@ -77,36 +78,22 @@ class StatusTest : public testing::TestWithParam<CppStacktracesMode> {
 
+// Catches the c10 error, and transforms it into a runtime error, calling
+// what() or what_without_backtrace() depending on IsShowCppStacktracesMode().
+#define XLA_THROW_RUNTIME_ERROR_FROM_C10_ERROR(block) \
```
Remove this? Prefer straightforward test code that just catches the c10::Error and verifies it.

zhanyong-wan reviewed (changes requested):
Thanks!

ysiraichi commented:
Here's a quick summary of the last commits:

- Removed `XLA_THROW_RUNTIME_ERROR_FROM_C10_ERROR` macro in favor of clearer `try-catch` tests
- Renamed `testing` namespace to `cpp_test`, following existing tests
